Bump minimum required boost version in master to 1.75.0

FranciscoPombal opened the pull request:
Allows for some `#if` cleanups + some cleanups in the build scripts; this is sort of a follow-up to d3216f63aeea558ed54c8bcb2dc9ab112db46577.

Let me know if there is anything wrong with the changes to the Jamfile, I'm not to familiar.

Also, I'm a little unsure about the change from `m_read_buffer.consume(m_read_buffer.size());` to `m_read_buffer.clear();` in `src/websocket_tracker_connection.cpp`.The docs for the former and latter method read, respectively (emphasis mine):

>Remove bytes from beginning of the **readable bytes**.
Removes n bytes from the beginning of the readable bytes.
All buffers sequences previously obtained using
`data` or `prepare` become invalid.
>
>`n`: The number of bytes to remove. If this number is greater than the number of readable bytes, all readable bytes are removed

>Set the size of the **readable and writable bytes** to zero.
This clears the buffer without changing capacity.
Buffer sequences previously obtained using `data` or
`prepare` become invalid.
>
>No-throw guarantee.

I did not study the code in detail, so I don't know if this makes any difference. But from a cursory look, it makes sense and seems to be OK, since in the libtorrent code I only see this buffer being read from, not written to, so the part about "and writable..." in the documentation should have no effect on the current behavior.

arvidn commented on include/libtorrent/aux_/utp_stream.hpp:
```diff
@@ -434,7 +434,6 @@ struct TORRENT_EXTRA_EXPORT utp_stream
 	}
 
-#if BOOST_VERSION >= 106600
 	// Compatiblity with the async_wait method introduced in boost 1.66
```
this comment should be removed too

FranciscoPombal replied:
fixed.

arvidn commented on bindings/python/Jamfile:
```diff
@@ -33,41 +33,11 @@ if $(LIBTORRENT_PYTHON_INTERPRETER)
-	{
-		alias boost_python3 : boost_python ;
-	}
+	alias boost_python3 : boost_python ;
```
I think the `boost_python3` target can be removed altogether with this change. I'll try out an update to this PR

aldenml commented:
It's my impression that for some platforms and the most common way to install boost, the version could be behind one or two. For example, today `brew info boost` reports `1.74.0`, some of the CI taks are green just because `webtorrent` is `off`. Maybe is better to wait a few months.

Btw, this should be removed from the root `Jamfile`
```
<include>./deps/json/include
<define>BOOST_JSON_HEADER_ONLY
```

arvidn commented:
there should probably be a check in some header, causing an `#error` in case the boost version is too old. And as Alden points out, it might be tricky to get boost-1.75 on all CI just yet.

xavier2k6 commented:
Building docs need to be updated too to reflect this change.

FranciscoPombal commented:
@aldenml 

> It's my impression that for some platforms and the most common way to install boost, the version could be behind one or two. For example, today `brew info boost` reports `1.74.0`, some of the CI taks are green just because `webtorrent` is `off`. Maybe is better to wait a few months.

Hmm, I see that historically, Homebrew apparently takes a bit longer than I expected to include newer boost versions. Let's see what happens.

I've tested with CMake and `-Dwebtorrent=ON` on Ubuntu 18.04 and 20.04 thus far. I see that CI for Linux currently uses boost from the repos, but it shouldn't be too difficult to set up boost from upstream into `/opt/boost-x.xx.xx` (some sort of caching should be used in order to not incur the delay of building it every time). This is what I do locally and it "just works" because `/opt` is one of the locations that CMake searches by default when trying to find a suitable version of a library. Hopefully the same can work with BBv2 without much hassle.

> Btw, this should be removed from the root `Jamfile`
> 
> ```
> <include>./deps/json/include
> <define>BOOST_JSON_HEADER_ONLY
> ```

Thanks, fixed.

FranciscoPombal commented:
> Building docs need to be updated too to reflect this change.

Thanks, fixed. Updated remaining docs + bumped boost on appveyor (1.75 does not exist there yet, so it's obviously not expected to pass).

FranciscoPombal commented:
@arvidn 

> there should probably be a check in some header, causing an `#error` in case the boost version is too old.

This is already done in this PR in CMake. Tips on how to do the same in BBv2 are welcome.

arvidn commented:
>> there should probably be a check in some header, causing an #error in case the boost version is too old.
> 
> This is already done in this PR in CMake. Tips on how to do the same in BBv2 are welcome.

I was imagining a check in some header file. It would be more reliable and portable

FranciscoPombal commented:
@arvidn 

> I was imagining a check in some header file. It would be more reliable and portable

By having the check in the build system, you allow the build to fail the fastest in case the necessary minimum requirements are not met (see the logs of the failed CMake builds of this PR as an example). By putting the check only where it's used, you have no real guarantees of how much time will be wasted running a build that will eventually fail anyway (okay, admittedly this should always be bounded by the total time taken by the preprocessing stage, so it can't be _that_ bad).
Otherwise, you have to either include that check everywhere or go out of your way to make sure that hypothetical header file is the first to be included in the build process and depended on by everything else, at which point it's just like putting the check in the build system with extra steps. Furthermore, it can lead to undesirable consequences: off the top of my head, this would at the very least least force a full recompilation every time the minimum required dependency version is changed.

Do you expect people to build libtorrent by rolling their own build system, or "manually"? I guess someone would do that if they had a very specialized need - but in that case, I would expect them to have combed thoroughly through the documentation and probably the source code, so they would be aware of the minimum requirements anyway.

I'm curious about what you think of this. I think of the build system as a core part of the project - just like any other component, it has a set of design goals it must meet, a proper way it's meant to be used, separation of concerns relative to other components (for example, it's its job to ensure all dependencies are the appropriate versions); it can ossify, decay etc, so it must be tended to every once in a while so it doesn't get in the way in the future. It's part an parcel; as an additional example, I consider supporting more than one build system a commitment akin to supporting more than one crypto library, for example (I don't mean this is necessarily a bad thing).

From the assumption that build dependency version checking is a job for the buildsystem (as opposed to a job for the compiler's preprocessor, indirectly), saying one wants such a check to be "reliable an portable" is really saying that they want the build system to be "reliable and portable" in the first place. Of course, even if the buildsystem tool itself has such potential, it is always possible to use it in a very non-portable and unreliable way, much the same way it is possible to write very non-portable and unreliable C, for instance.

But I'm curious about other perspectives or compelling use cases that go against this "philosophy", or are otherwise not feasible to achieve simultaneously.

userdocs commented:
1_75 is going to report 1_74 until they fix this.

https://github.com/boostorg/boost/issues/446

Considering they use the bintray archive in the formula 

https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/boost.rb#L4

It might be confusing the issue.

FranciscoPombal commented:
Most things should pass now. But Appveyor is still only on Boost 1.73, and on Ubuntu 20.04 the latest available version is 1.71. Perhaps it would be better to always get the latest headers from source as it is already being done for some Windows jobs.

FranciscoPombal commented:
@arvidn To be honest, I don't know enough about b2 to make this work. Can you make a PR against this branch that fixes the CI?

xavier2k6 commented:
@arvidn now that Boost 1.77 is available in AppVeyor & there's a PR open to raise it to that, does this PR still have relevance/need to be reworked?!

(Also going forward, their plan is to have two latest versions available e.g. 1.77 & 1.78 when released etc etc.) 

Will `RC_2_0` & `RC_1_2` Minimum Boost requirements also be bumped?

arvidn commented:
> Will RC_2_0 & RC_1_2 Minimum Boost requirements also be bumped?

I wasn't going to do that until appveyor either drops support for the older images, or updates those images to only have newer boost versions.

arvidn commented:
I'm still hesitant to require a newer version of boost than what's available in a distro's package manager.
